PR: Add documentation for the Spyder-Unittest plugin (#272)
spyder-ide/spyder-docs: doc/plugins/unittest.rst

###############
Spyder Unittest
###############

> **Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder, allowing you to run test suites and view the results in the IDE.

Reviewer: Add link in unit test to useful docs. Maybe this one: https://docs.python.org/3/library/unittest.html but let me know @ccordoba12 if you have something better.

Reviewer 2: ```suggestion
**Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder, such as Pytest, Unittest and Nose, allowing you to run test suites and view the results in the IDE.
```

- I agree with @juanis2112 that it'd be a nice idea to add links for the testing frameworks we support.
- That's why I added an explicit mention to them above.

Author: Links to Pytest, Unittest and Nose was added.

.. image:: /images/unittest/unittest-run.png
   :alt: Spyder Unittest in Spyder

> Using unittest inside Spyder allows you to take advantage of their web interface alongside Spyder’s powerful features such as the Variable explorer, console and debugger.

Reviewer: ```suggestion
Using unittest inside Spyder allows you to take advantage of their web interface alongside Spyder’s powerful features such as the Variable explorer, console and debugger.
```
This is left from the notebook one and I'm not sure it applies to this one but @ccordoba12 can confirm.

Reviewer 2: Yeah, it doesn't apply here.

Author: this was removed

===================
Installing Unittest
===================

.. code-block:: bash

   conda install -c spyder-ide spyder-unittest

.. important::

>    Please make sure you have Python's pytest package installed if you are going to use that program as well.

Reviewer 2: ```suggestion
   Please make sure you have Python's Pytest package installed if you are going to use that program as well.
```

Small improvement.

Author: changed case of letter to upper.

.. code-block:: bash

   conda install pytest

Restart Spyder in order to be able to use the plugin.

==============
Using Unittest
==============

> When Unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.

Reviewer 2: ```suggestion
When Spyder-unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.
```

Unittest is the name of standard library module for testing, so I think it's better to be more explicit here.

Author: changed.

.. image:: /images/unittest/unittest-view-panes.png
   :alt: Spyder showing view panes Unittest

> You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots views.

Reviewer 2: ```suggestion
You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots panes.
```

Small correction.

Author: changed

> Press on the  :guilabel:`Run Green Arrow` to run the testing.

Reviewer: ```suggestion
Press on the  :guilabel:`Run Green Arrow` to run the testing.
```
I think here it might be worth to explain that we need to have some tests written in the editor and maybe mention what a test is in Python with a link to this page where there is more information about it https://docs.python.org/3/library/unittest.html. What do you think @ccordoba12 ?

Reviewer 2: Yeah, I think that's important to explain first, i.e. before clicking `Run`, you need to have at least a file whose name starts by `test_` and then functions on it that also start by `test_`. And below that explanation, we could also show a screenshot showing that in the editor, as @juanis2112 suggested.

Of course, that's if we're going to use Pytest to show the plugin functionality, which I think is much simpler to explain than Unittest.

Author: The link to the Pytest website and the inclusion of the requirements for a testing file to run were included.

.. image:: /images/unittest/unittest-run-button.png
   :alt: Spyder showing Unittest run button

You can get an expanded view of the testing messages by clicking on the expand option.

.. image:: /images/unittest/unittest-expanded-message-view.png
   :alt: Spyder showing Unittest expanded message view

> Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.

Reviewer: Before this one I think it could be worth adding the "Advanced options" since these options are not that common.

Reviewer 2: Agreed.

Author: Added information on the configuration view, dependencies view, and the output viewer.

.. image:: /images/unittest/unittest-hamburger-menu.png
   :alt: Spyder Unittest hamburger menu in Spyder

The configuration view will show you how to choose the pytest option and the directory of where you want to run the test file(s) from.

.. image:: /images/unittest/unittest-configuration-view.png
   :alt: Spyder Unittest configuration view in Spyder

> The dependencies view will show you all of your installed package versions.

Reviewer 2: ```suggestion
The dependencies view will show you all of your installed testing frameworks.
```

Small correction.

Author: changed.

.. image:: /images/unittest/unittest-dependencies-view.png
   :alt: Spyder Unittest dependencies view in Spyder

> The output viewer can be used to directly view the output results of your testing.  This view can be very useful for viewing the results of unittests.

Reviewer: ```suggestion
Finally, the output viewer can be used to directly view the output results of your testing.  This view can be very useful for viewing the results of unittests.
```
To improve the flow

Author: changed to suggested text.

###############
Spyder Unittest
###############

**Spyder-unittest** is a plugin that integrates popular unit test frameworks with Spyder: such as Pytest, Unittest and Nose. It allows you to run test suites and view the results in the IDE.

.. image:: /images/unittest/unittest-run.png
   :alt: Spyder Unittest in Spyder

| `Pytest <https://docs.pytest.org/en/stable/getting-started.html#install-pytest>`_

> | `Unittest <https://docs.python.org/3/library/unittest.html#module-unittest>`_

Reviewer 3: I believe this was superceded by another suggestion of yours.

| `Nose <https://nose.readthedocs.io/en/latest/>`_

===================
Installing Unittest
===================

.. code-block:: bash

   conda install -c spyder-ide spyder-unittest

.. important::

   Please make sure you have Python's Pytest package installed if you are going to use that program as well.

.. code-block:: bash

   conda install pytest

Restart Spyder in order to be able to use the plugin.

==============
Using Unittest
==============

When Spyder-unittest is installed, it will be available under the menu item :menuselection:`View --> Panes --> Unit testing`.

.. image:: /images/unittest/unittest-view-panes.png
   :alt: Spyder showing view panes Unittest

You will see it then as a tab in the bottom of the top right area in the same location as the Variable explorer, Help, and Plots panes.

> Press on the  :guilabel:`Run Green Arrow` to run the testing.

Reviewer: ```suggestion
Press on the  :guilabel:`Run Green Arrow` to run the testing.

.. image:: /images/unittest/unittest-hamburger-menu.png
   :alt: Spyder showing Unittest run button

```

I think the screenshot makes more sense here.

Pytest will run all files of the form :guilabel:`test_*.py` or :guilabel:`*_test.py` in the current directory and its subdirectories.

For more on the Python Pytest package:

| `Pytest Run Guide <https://docs.pytest.org/en/stable/getting-started.html#run-multiple-tests>`_

| `Real Python Article on using Pytest <https://realpython.com/pytest-python-testing/>`_

| `Software Testing Help's Pytest Tutorial <https://www.softwaretestinghelp.com/pytest-tutorial/>`_

> .. image:: /images/unittest/unittest-run-button.png

Reviewer: ```suggestion
```

>    :alt: Spyder showing Unittest run button

Reviewer: ```suggestion
```

Reviewer 3: Looks like this was fixed elsewhere

You can get an expanded view of the testing messages by clicking on the expand option.

.. image:: /images/unittest/unittest-expanded-message-view.png
   :alt: Spyder showing Unittest expanded message view

> Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.

Reviewer: ```suggestion

==================
Additional Options
==================

Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.
```

Adding a title to the Options menu section.

Pytest will run all files of the form :guilabel:`test_*.py` or :guilabel:`*_test.py` in the current directory and its subdirectories.

For more on the Python Pytest package:

| `Pytest Run Guide <https://docs.pytest.org/en/stable/getting-started.html#run-multiple-tests>`_

| `Real Python Article on using Pytest <https://realpython.com/pytest-python-testing/>`_

| `Software Testing Help's Pytest Tutorial <https://www.softwaretestinghelp.com/pytest-tutorial/>`_

You can get an expanded view of the testing messages by clicking on the expand option.

.. image:: /images/unittest/unittest-expanded-message-view.png
   :alt: Spyder showing Unittest expanded message view

==================
Additional Options
==================

Press on the  :guilabel:`Hamburger Menu Icon` to view different unittest options.

.. image:: /images/unittest/unittest-hamburger-menu.png
   :alt: Spyder Unittest hamburger menu in Spyder

The configuration view will show you how to choose the pytest option and the directory of where you want to run the test file(s) from.

.. image:: /images/unittest/unittest-configuration-view.png
   :alt: Spyder Unittest configuration view in Spyder

> The output viewer can be used to directly view the output results of your testing.  This view can be very useful for viewing the results of unittests.

Reviewer 3: ```suggestion
The :guilabel:`Show output` dialog can be used to directly view the raw output of your tests.
This can be useful to see more details of the problem when a test fails.
```

* Rewrite redundant second sentence to avoid duplicating the content in the first and instead provide unique information
* Use more specific phrasing and teminology
* Avoid word repetition
* Refer to the actual labels shown in the GUI and use appropriate roles
* One sentence per line

.. image:: /images/unittest/unittest-output-viewer.png
   :alt: Spyder Unittest output viewer in Spyder

> Finally, the dependencies view will show you all of your installed testing frameworks.

Reviewer 3: ```suggestion
Finally, the :guilabel:`Dependencies` dialog will show all of your installed testing frameworks and their versions.
```

* Use more specific phrasing and teminology
* Refer to the actual labels shown in the GUI and use appropriate roles
